### PR TITLE
GaussianBlur CV-CUDA Backend

### PR DESCRIPTION
**Pull request by justincdavis**
### Summary

Implement the CV-CUDA backend kernel for gaussian_blur

### How to use

```
import cvcuda
import torchvision.transforms.v2.functional as F

cv_tensor = cvcuda.Tensor((1, 224, 224, 3), cvcuda.Type.U8, cvcuda.TensorLayout.NHWC)
# dispatched to F.gaussian_blur_cvcuda
blurred = F.gaussian_blur(cv_tensor, (5, 5))
```

### CODE REVIEW

File: torchvision/transforms/v2/functional/_misc.py [hunk 1]
```diff
@@ -99,6 +106,41 @@ def _get_gaussian_kernel2d(
     return kernel2d
 
 
+def _validate_kernel_size_and_sigma(
```
Comment (AntoineSimoulin) [inline torchvision/transforms/v2/functional/_misc.py:107]: Should we use `_validate_kernel_size_and_sigma` in `gaussian_blur_image` since we extracted the logic from here?

File: torchvision/transforms/v2/functional/_misc.py [hunk 2]
```diff
@@ -181,6 +223,27 @@ def gaussian_blur_video(
...
+    return cvcuda.gaussian(
+        image,
+        tuple(kernel_size),
+        tuple(sigma),
+        border=cvcuda.Border.REFLECT,
+    )
+
+
+if CVCUDA_AVAILABLE:
+    _gaussian_blur_cvcuda_registered = _register_kernel_internal(gaussian_blur, _import_cvcuda().Tensor)(
```
Comment (AntoineSimoulin) [inline torchvision/transforms/v2/functional/_misc.py]: We should be able to remove `_gaussian_blur_cvcuda_registered = ` and only keep `_register_kernel_internal(gaussian_blur, _import_cvcuda().Tensor)(_gaussian_blur_cvcuda)`.

File: test/test_transforms_v2.py [hunk 3]
```diff
@@ -3884,7 +3884,15 @@ def test_kernel_video(self):
```
Comment (AntoineSimoulin) [inline test/test_transforms_v2.py:3928]: Should we add a test to check the cvcuda kernel?

File: test/test_transforms_v2.py [hunk 4]
```diff
@@ -3994,6 +4017,39 @@ def test_functional_image_correctness(self, dimensions, kernel_size, sigma, dtyp
...
+            ((10, 12), (3, 3), 0.5),
+            ((10, 12), (3, 5), 0.8),
+            ((10, 12), (3, 5), 0.5),
+            ((26, 28), (23, 23), 1.7),
+        ],
+    )
+    @pytest.mark.parametrize("color_space", ["RGB", "GRAY"])
+    @pytest.mark.parametrize("batch_dims", [(1,), (2,), (4,)])
+    @pytest.mark.parametrize("dtype", [torch.uint8, torch.float32])
+    def test_functional_cvcuda_parity(self, dimensions, kernel_size, sigma, color_space, batch_dims, dtype):
```
Comment (AntoineSimoulin) [inline test/test_transforms_v2.py]: Can we rename `test_functional_cvcuda_correctness` for consistency?

File: test/test_transforms_v2.py [hunk 5]
```diff
@@ -3994,6 +4017,39 @@ def test_functional_image_correctness(self, dimensions, kernel_size, sigma, dtyp
...
+            size=(height, width), color_space=color_space, batch_dims=batch_dims, dtype=dtype, device="cuda"
+        )
+        image_cvcuda = F.to_cvcuda_tensor(image_tensor)
+
+        expected = F.gaussian_blur_image(image_tensor, kernel_size=kernel_size, sigma=sigma)
+        actual = F._misc._gaussian_blur_cvcuda(image_cvcuda, kernel_size=kernel_size, sigma=sigma)
+        actual_torch = F.cvcuda_to_tensor(actual)
+
+        if dtype.is_floating_point:
+            torch.testing.assert_close(actual_torch, expected, rtol=0, atol=0.3)
```
Comment (AntoineSimoulin) [inline test/test_transforms_v2.py]: Why setting `atol=0.3` here?

**Review by AntoineSimoulin — COMMENTED** [review]
Thanks for the PR @justincdavis. Left a few comments, looking good otherwise!

File: torchvision/transforms/v2/functional/_misc.py [hunk 6]
```diff
@@ -99,6 +106,41 @@ def _get_gaussian_kernel2d(
...
+        elif isinstance(sigma, (int, float)):
+            s = float(sigma)
+            sigma = [s, s]
+        else:
+            raise TypeError(f"sigma should be either float or sequence of floats. Got {type(sigma)}")
+    for s in sigma:
+        if s <= 0.0:
+            raise ValueError(f"sigma should have positive values. Got {sigma}")
+
+    return kernel_size, sigma
```
Comment (AntoineSimoulin) [inline torchvision/transforms/v2/functional/_misc.py:139]: Can we make sure we address this lint error? 
```python
Incompatible return
  value type (got
  "tuple[Sequence[int], list[float] | list[float | int] | list[Any] | tuple[Any, ...]]",
  expected "tuple[list[int], list[float]]")  [return-value]
          return kernel_size, sigma
```

File: torchvision/transforms/v2/functional/_misc.py (hunk 1 above)
Comment (justincdavis) [reply torchvision/transforms/v2/functional/_misc.py:107]: Yes! I have refactored accordingly, thanks for pointing that out

File: test/test_transforms_v2.py (hunk 5 above)
Comment (justincdavis) [reply test/test_transforms_v2.py]: Good question! I added a comment on `atol=0.3`, most likely from floating point differences between the underlying `filter2d` in CV-CUDA compared to `torch.conv2d`. Let me know if you want more explanation and/or something else here.

File: torchvision/transforms/v2/functional/_misc.py (hunk 6 above)
Comment (justincdavis) [reply torchvision/transforms/v2/functional/_misc.py:139]: This should be addressed, don't see any lint errors locally, thanks for pointing this out!

File: test/test_transforms_v2.py (hunk 3 above)
Comment (justincdavis) [reply test/test_transforms_v2.py:3928]: Resolving from [comment chain here](https://github.com/pytorch/vision/pull/9278#discussion_r2556628689)

File: test/test_transforms_v2.py (hunk 5 above)
Comment (AntoineSimoulin) [reply test/test_transforms_v2.py]: Should we set it as in `test_functional_image_correctness` with `torch.testing.assert_close(actual, expected, rtol=0, atol=1)` for consistency?

Comment (justincdavis) [reply test/test_transforms_v2.py]: @AntoineSimoulin I ended up rewriting the test setup, and moved all the tests into a single block. Both CV-CUDA and torchvision share the same assert statement now. LMK if you think it looks like a good change.

File: test/test_transforms_v2.py [hunk 7]
```diff
@@ -4018,11 +4043,22 @@ def test_make_params(self, sigma):
...
+            pytest.param(
+                "cvcuda.Tensor", marks=pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="CVCUDA not available")
+            ),
+        ],
+    )
+    def test_functional_image_correctness(self, dimensions, kernel_size, sigma, dtype, device, input_type):
         if dtype is torch.float16 and device == "cpu":
             pytest.skip("The CPU implementation of float16 on CPU differs from opencv")
+        if (dtype != torch.float32 and dtype != torch.uint8) and input_type == "cvcuda.Tensor":
+            pytest.skip("CVCUDA does not support non-float32 or uint8 dtypes for gaussian blur")
```
Comment (zy1git) [inline test/test_transforms_v2.py]: I feel that this comment is bit confusing:

- Does it mean "non-(float32 or uint8)" → neither float32 nor uint8?

- Or "(non-float32) or uint8" → something else entirely?

Thus, I recommend to use "CVCUDA only supports float32 and uint8 dtypes for gaussian blur".

File: test/test_transforms_v2.py [hunk 8]
```diff
@@ -4042,9 +4078,17 @@ def test_functional_image_correctness(self, dimensions, kernel_size, sigma, dtyp
...
+        if input_type == "cvcuda.Tensor":
+            image = image.unsqueeze(0)
+            image = F.to_cvcuda_tensor(image)
 
-        torch.testing.assert_close(actual, expected, rtol=0, atol=1)
+        actual = F.gaussian_blur(image, kernel_size=kernel_size, sigma=sigma)
+
+        if input_type == "cvcuda.Tensor":
+            actual = F.cvcuda_to_tensor(actual)
+            actual = actual.squeeze(0).to(device=device)
```
Comment (zy1git) [inline test/test_transforms_v2.py]: We can also use `actual=actual[0].to(device=device)` since batch size is guaranteed to be 1 in this case. Not sure we need to be consistent to the implementation here: https://github.com/pytorch/vision/pull/9277/changes#diff-9c2dde92db86c123fee225e39b7c1ef96e08a3e79a9dcc9a2d68b21ed51a81d0R1315

Comment (zy1git) [inline test/test_transforms_v2.py]: I think we can also use `actual = actual[0].to(device=device)` since the batch size is always 1 in this case. Not sure we need to be consistent with the implementation here: https://github.com/pytorch/vision/pull/9277/changes#diff-9c2dde92db86c123fee225e39b7c1ef96e08a3e79a9dcc9a2d68b21ed51a81d0R1315

File: test/test_transforms_v2.py [hunk 9]
```diff
@@ -3936,7 +3936,15 @@ def test_kernel_video(self):
...
     @pytest.mark.parametrize(
         "make_input",
-        [make_image_tensor, make_image_pil, make_image, make_video],
+        [
+            make_image_tensor,
+            make_image_pil,
+            make_image,
+            make_video,
+            pytest.param(
+                make_image_cvcuda, marks=pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="CVCUDA is not available")
```
Comment (zy1git) [inline test/test_transforms_v2.py]: See this PR: https://github.com/pytorch/vision/pull/9305/changes

There are other parts with similar issues that also need to be addressed.

**Review by zy1git — COMMENTED** [review]
I’ve left some comments on this PR.
Please feel free to address them or reach out if you’d like to discuss any points further.

File: test/test_transforms_v2.py [hunk 10]
```diff
@@ -4034,9 +4064,17 @@ def test_functional_image_correctness(self, dimensions, kernel_size, sigma, dtyp
...
+        if input_type == "cvcuda.Tensor":
+            image = image.unsqueeze(0)
+            image = F.to_cvcuda_tensor(image)
 
-        torch.testing.assert_close(actual, expected, rtol=0, atol=1)
+        actual = F.gaussian_blur(image, kernel_size=kernel_size, sigma=sigma)
+
+        if input_type == "cvcuda.Tensor":
+            actual = F.cvcuda_to_tensor(actual)
+            actual = actual[0].to(device=device)
```
Comment (NicolasHug) [inline test/test_transforms_v2.py]: Nit: this would be slightly easier to read as a one-liner


```suggestion
            actual = F.cvcuda_to_tensor(actual)[0].to(device)
```

File: test/test_transforms_v2.py [hunk 11]
```diff
@@ -4010,11 +4031,20 @@ def test_make_params(self, sigma):
...
+        [
+            tv_tensors.Image,
+            pytest.param("cvcuda.Tensor", marks=pytest.mark.needs_cvcuda),
+        ],
+    )
+    def test_functional_image_correctness(self, dimensions, kernel_size, sigma, dtype, device, input_type):
         if dtype is torch.float16 and device == "cpu":
             pytest.skip("The CPU implementation of float16 on CPU differs from opencv")
+        if not (dtype == torch.float32 or dtype == torch.uint8) and input_type == "cvcuda.Tensor":
+            pytest.skip("CV-CUDA only supports float32 and uint8 dtypes for gaussian blur")
```
Comment (NicolasHug) [inline test/test_transforms_v2.py]: TODO: need to check where this is validated - what's the error message? We should add a test for that, and this should probably be xfail.

File: test/test_transforms_v2.py
Comment (NicolasHug) [inline test/test_transforms_v2.py:1]: Note for self: need to check that all the relevant tests have been properly parametrized.

File: torchvision/transforms/v2/functional/_misc.py [hunk 12]
```diff
@@ -181,6 +195,25 @@ def gaussian_blur_video(
     return gaussian_blur_image(video, kernel_size, sigma)
 
 
+def _gaussian_blur_image_cvcuda(
+    image: "cvcuda.Tensor", kernel_size: list[int], sigma: Optional[list[float]] = None
+) -> "cvcuda.Tensor":
+    cvcuda = _import_cvcuda()
+
```
Comment (zy1git) [inline torchvision/transforms/v2/functional/_misc.py:202]: Do we need to add the dtype checking here in the `Gaussianblur `implementation like the `normalize` implementation in:https://github.com/pytorch/vision/pull/9279/changes#diff-9d8095db7c452f3338f608726a3aa45e0f7956a79be95c101b3e0c3c4ea7ece4R93

Like:

```python
if image.dtype not in (cvcuda.Type.U8, cvcuda.Type.F32):
        raise ValueError(
            f"CV-CUDA gaussian blur only supports uint8 and float32 tensors. Got {image.dtype}."
        )
```

File: test/test_transforms_v2.py (hunk 11 above)
Comment (zy1git) [reply test/test_transforms_v2.py]: I changed the `skip `to `xfail `in the latest commit. 

I am a bit confused with an extra test to add for that. Do you mean the test similar to the test I added here: https://github.com/pytorch/vision/pull/9279/changes#diff-9c2dde92db86c123fee225e39b7c1ef96e08a3e79a9dcc9a2d68b21ed51a81d0R5606

If so, then we need to add an explicit dtype validation to gaussian blur as I commented here: https://github.com/pytorch/vision/pull/9280/changes#r2679384750

Please let me know if there is any misunderstanding.

File: torchvision/transforms/v2/functional/_misc.py (hunk 12 above)
Comment (NicolasHug) [reply torchvision/transforms/v2/functional/_misc.py:202]: These checks might exist internally within `cvcuda.gaussian`, we should check. If they exist already, we don't need to add them ourselves within torchvision. If they do not exist then this questions the need for the previous `pytest.skip()` (now `pytest.xfail`)

File: test/test_transforms_v2.py (hunk 11 above)
Comment (NicolasHug) [reply test/test_transforms_v2.py]: > Do you mean the test similar to the test I added here: [#9279 (changes)](https://github.com/pytorch/vision/pull/9279/changes#diff-9c2dde92db86c123fee225e39b7c1ef96e08a3e79a9dcc9a2d68b21ed51a81d0R5606)

yes, that's what I meant

File: test/test_transforms_v2.py [hunk 13]
```diff
@@ -3926,9 +3926,20 @@ def test_kernel_image_errors(self):
     def test_kernel_video(self):
         check_kernel(F.gaussian_blur_video, make_video(), kernel_size=(3, 3))
 
+    @needs_cvcuda
+    def test_functional_error_cvcuda(self):
+        with pytest.raises(RuntimeError, match="INVALID_DATA_TYPE"):
+            F.gaussian_blur(make_image_cvcuda(dtype=torch.float64), kernel_size=(3, 3), sigma=0.8)
```
Comment (NicolasHug) [inline test/test_transforms_v2.py]: Thanks for adding the test! From your [doc](https://fburl.com/gdoc/l9ggyzfb), the error message that is raised by CV-CUDA doesn't seem to be very detailed:

> NVCV_ERROR_INVALID_ARGUMENT: :0 (INVALID_DATA_TYPE)”

This says that the dtype is invalid, but it doesn't say which types are valid, so it's not super useful to the user. So raise our own `ValueError` exception in the kernel, so that we can say something a bit better, like

> CV-CUDA's GaussianBlur only supports float32 and uint8 dtype, got <the_dtype_here>

Comment (zy1git) [reply test/test_transforms_v2.py]: Thanks a lot for the feedback. I have fixed the error message accordingly by adding our own `ValueError `exception in the kernel.
